File: onyx-database/src/main/java/com/onyx/diskmap/node/SkipListNode.java
package com.onyx.diskmap.node;

import com.onyx.buffer.BufferStream;
import com.onyx.exception.BufferingException;

/**
 * Created by tosborn1 on 1/6/17.
 *
 * This is a record pointer for a skip list.
 */
public class SkipListNode<K> extends SkipListHeadNode {

    public static final int BASE_SKIP_LIST_NODE_SIZE = Long.BYTES * 4 + Byte.BYTES + Integer.BYTES;
    public long recordPosition;
    public long recordId;
    public int recordSize;
    public K key;

    @SuppressWarnings("unused")
    public SkipListNode()
    {

    }

    public SkipListNode(K key, long position, long recordPosition, byte level, long next, long down, int recordSize, long recordId) {
        this.position = position;
        this.recordPosition = recordPosition;
        this.level = level;
        this.next = next;
        this.down = down;
        this.key = key;
        this.recordSize = recordSize;
        this.recordId = recordId;
    }

    @Override
    public void read(BufferStream buffer) throws BufferingException {
        recordPosition = buffer.getLong();
        recordSize = buffer.getInt();
        super.read(buffer);
        recordId = buffer.getLong();
        key = (K)buffer.getValue();
    }

    @Override
    public void write(BufferStream buffer) throws BufferingException {
        buffer.putLong(recordPosition);
        buffer.putInt(recordSize);
        super.write(buffer);
        buffer.putLong(recordId);
        buffer.putObject(key);
    }

}

File: onyx-database/src/main/kotlin/com/onyx/diskmap/data/SkipListNode.kt
package com.onyx.diskmap.data

import com.onyx.buffer.BufferStream
import com.onyx.exception.BufferingException

/**
 * Created by Tim Osborn on 1/6/17.
 *
 * This is a record pointer for a skip list.
 */
@Suppress("UNCHECKED_CAST")
class SkipListNode<K> : SkipListHeadNode {
    var recordPosition: Long = 0
    var recordId: Long = 0
    var key: K = KEY as K

    constructor()

    constructor(key: K, position: Long, recordPosition: Long, level: Byte, next: Long, down: Long, recordId:Long) {
        this.position = position
        this.recordPosition = recordPosition
        this.level = level
        this.next = next
        this.down = down
        this.key = key
        this.recordId = recordId
    }

    @Throws(BufferingException::class)
    override fun read(buffer: BufferStream) {
        super.read(buffer)
        recordPosition = buffer.long
        recordId = buffer.long
        key = buffer.value as K
    }

    @Throws(BufferingException::class)
    override fun write(buffer: BufferStream) {
        super.write(buffer)
        buffer.putLong(recordPosition)
        buffer.putLong(recordId)
        buffer.putObject(key)
    }

    companion object {
        val BASE_SKIP_LIST_NODE_SIZE = java.lang.Long.BYTES * 2 + java.lang.Byte.BYTES
        val SKIP_LIST_NODE_SIZE = java.lang.Long.BYTES * 4 + java.lang.Byte.BYTES
        object KEY
    }

}
